Objective-C/CBLDictionary.mm

```diff
@@ -264,28 +301,42 @@ - (BOOL) isEqual: (id)object {
     if (self.count != other.count)
         return NO;
     
-    for (MDict<id>::iterator i(_dict); i; ++i) {
-        NSString* key = i.nativeKey();
-        id value = i.nativeValue();
-        if (value) {
-            if (![value isEqual: [other valueForKey: key]])
-                return NO;
-        } else {
-            if ([other valueForKey: key] || ![other containsValueForKey: key])
-                return NO;
+    CBL_LOCK(self.sharedLock) {
+        for (MDict<id>::iterator i(_dict); i; ++i) {
+            NSString* key = i.nativeKey();
+            id value = i.nativeValue();
+            if (value) {
+                if (![value isEqual: [other valueForKey: key]])
```

Comment:
Calling `other` while locked is unsafe; it can deadlock if another thread is calling `isEqual:` on `other`. (Assuming that the sharedLock is `self`, not `db`.)